core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -13,22 +13,55 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
+import java.util.function.Supplier;
 
 public class ThreadContextClassLoader
-        implements Closeable
 {
-    private final ClassLoader originalThreadContextClassLoader;
+    private ThreadContextClassLoader() {}
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
+    public static <T> T withClassLoader(ClassLoader newThreadContextClassLoader, Supplier<T> supplier)
     {
-        this.originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
         Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
     }
 
-    @Override
-    public void close()
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(newThreadContextClassLoader, (Supplier<T>) supplier);
```

Comment:
i would delegate the other way around.

Reply:
I was thinking of delegation from more-specific `ThrowingSupplier<T, E>`-version to more-generic `Supplier<T>`-version to keep `Supplier<T>`-version unaware of any checked exceptions and to point to proper `Supplier::get`, which is overridden in `ThrowingSupplier`.

Reply:
done